### PR TITLE
File editor improvements

### PR DESCRIPTION
**Pull request by mausworks**
## Description

This PR refactors components around the `FileTree` and the `FileTree` itself. Reducing complexity and fixing a few bugs in the process.

- Fixed: You're no longer allowed to change the extension of `.orchest` files
- Fixed: When a file is renamed to `.orchest` and there are no pipelines available, it is automatically loaded.
- Improves: Clicking outside the rename text box now saves the rename (like it does in Finder & VSCode)
- Improves: Overwriting files now always require confirmation
- Improves: Renaming is now optimistic, and feels more snappy.

Fixes: #1019 

### Requirements

When doing the refactoring around moving files, these are the requirements that I have used: they reflect the current behavior of the file browser, but includes some (reasonable) additions.

Before the move …

- If any of the moved files are “locked” because of one or more active sessions, the move is aborted, and a dialog with a button to stop the ongoing sessions is shown.
- If any of the moved files would overwrite an existing file, the move is aborted and a dialog explaining which files would be overwritten is shown.
- If neither of the previous dialogs were shown and one or more files are being moved to a different directory, a dialog asking the user to confirm the move is shown.
- If the result of any move operation will result in the loss of functionality (i.e. “bricking something”), a dialog asking the user to confirm the move is shown.
- Otherwise: if a single file is being renamed (i.e. ”is moved within the same directory”) the move proceeds without confirmation.

Performing the move …

- If a pipeline file is moved within the same project, the “rename pipeline endpoint” should be called.
- Otherwise, the “rename file” endpoint should be called.

After the move …

- The file tree should be reloaded
- If one or more pipeline files were moved, all pipelines should be re-fetched
- The moved files are selected at their new location

### Video

Here's me clicking around a bit:

https://user-images.githubusercontent.com/8259221/179518363-0b06d782-106d-40b5-836e-c01238bef9df.mov

## Checklist

- [x] I have manually tested my changes and I am happy with the result.
- [x] The documentation reflects the changes.
- [x] The PR branch is set up to merge into `dev` instead of `master`.
- [x] I haven't introduced breaking changes that would disrupt existing jobs, i.e. backwards compatibility is maintained.

### CODE REVIEW

**Review by yannickperrenet — APPROVED** [review]
Absolutely love the elaborate PR description. Makes it a lot easier to give a high-level review based on the logic behind the implementation (e.g. cancel rename if it would overwrite an existing file). :100: 

Works like a charm as well :)

----

I do have a nitpick though. How about making the rename functionality similar to VSCode:

https://user-images.githubusercontent.com/26223174/179228549-c57af2c3-1c4e-4cc7-b028-3b7f1b192c69.mp4

Makes it easier to rename the file. Given that we would also start preventing a rename of the `.orchest` extension, we might as well not put the cursor at the end of the extension.

**Comment by ncspost** [comment]
> Fixed: Clicking outside the rename text box now cancels the rename

Could we change this to enters the rename please?

**Comment by ncspost** [comment]
image

Title: File type warning
Body: `.ipynb` and `.orchest` file types can't be opened within `/data`. Would you still like move this file?
Buttons: Cancel + Move files

**Comment by ncspost** [comment]
CleanShot 2022-07-15 at 14 39 25@2x

I just created a ipynb file without an existing pipeline and renamed it to orchest. It seems a bit confused...

CleanShot 2022-07-15 at 14 40 05@2x

This is when I double-click the file.

**Comment by ncspost** [comment]
Great work @mausworks - I've left a few comments. 

Just a general theme: we have lots of dialogs in the product and I'm trying to reduce that number (https://linear.app/orchest/issue/ORC-669/too-many-dialogs). 

Consider using snack bars or banners/alerts to convey system feedback less intrusively. Use dialogs sparingly.

CleanShot 2022-07-15 at 14 46 27@2x

https://material.io/components/dialogs#usage

**Comment by mausworks** [comment]
I agree that some of these dialogs should be converted to toasts/snackbar messages. But I don't think we have anything set up for those yet.

For now, I'll leave them as dialogs, and then see if we can do a general purpose overhaul once we have abstractions in place for them. 👌

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx [hunk 1]
```diff
@@ -244,405 +214,233 @@ export const FileTree = React.memo(function FileTreeComponent({
...
+              await Promise.all(
+                lockedFiles.map(({ pipelineUuid }) => stopSession(pipelineUuid))
+              );
+              resolve(true);
+              return true;
+            },
+          }
+        );
+      } else {
+        const willBrickSomeFile =
```
Comment (iannbing) [inline services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx:362]: [nit] Not sure if `brick` is a commonly-known verb, but it did make me think for a couple of seconds. Perhaps a boring "block" will also do? And "file" should be plural?

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx [hunk 2]
```diff
@@ -5,162 +5,122 @@ import { useSessionsContext } from "@/contexts/SessionsContext";
 import { useCustomRoute } from "@/hooks/useCustomRoute";
 import { fetchPipelines } from "@/hooks/useFetchPipelines";
```
Comment (iannbing) [inline services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx:6]: There are some eslint warning about [react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920). Could you go through them? If they are intentional, please comment them with `// eslint-disable-line react-hooks/exhaustive-deps` and the reason (if not obvious). 🙂

**Review by iannbing — COMMENTED** [review]
Great work! Superb cleanup and great improvements 💯. I only have some minor comments.

File: services/orchest-webserver/client/src/hooks/useOnClickOutside.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    document.addEventListener("mousedown", listener);
+    document.addEventListener("touchstart", listener);
+
+    return () => {
+      document.removeEventListener("mousedown", listener);
+      document.removeEventListener("touchstart", listener);
+    };
+  }, [ref, handler]);
+};
```
[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/hooks/useOnClickOutside.ts:24]: How did I miss that? Definitely, thanks. 👍

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx [hunk 4]
```diff
@@ -244,405 +214,233 @@ export const FileTree = React.memo(function FileTreeComponent({
...
-          }
-        );
+        lockedFiles.push({ path, root, pipelineUuid });
       }
 
-      return [pipelinesWithSession.length === 0, pipelinesWithSession];
+      return lockedFiles;
     },
-    [getSession, projectUuid, setConfirm, stopSession, pipelineDics]
+    [getSession, pipelines]
```
[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx:258]: My eslint in VSCode seems to have broken, I'll do a pass. 👌

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx (hunk 1 above)
Comment (mausworks) [reply services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx:362]: I'm not a huge fan either, it definitely isn't the most commonly known concept. I'll reconsider the naming.

**Comment by mausworks** [comment]
@ncspost 

>Fixed: Clicking outside the rename text box now cancels the rename
>>Could we change this to enters the rename please?

So you mean it submits the rename? I'm

**Comment by mausworks** [comment]
> I just created a ipynb file without an existing pipeline and renamed it to orchest. It seems a bit confused... 
> ... this is when I double click the file

Should we maybe provide an action here called "Create pipeline"? Seems like a reasonable next step from this point.

**Comment by mausworks** [comment]
@ncspost 

- Title: File xyz already exists
- Body: Renaming `abc` to `xyz` will overwrite the existing file. Would you like to do this?
- Buttons: Cancel & Overwrite existing file

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileManager.tsx [hunk 5]
```diff
@@ -247,7 +247,7 @@ export function FileManager() {
     hasValue(fileTrees[root])
   );
 
-  const onRename = React.useCallback(
+  const onMoved = React.useCallback(
```
[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/pipeline-view/file-manager/FileManager.tsx:250]: The callback _is_ called after moving, so I guess the naming works as intended? 🤔

File: services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx [hunk 6]
```diff
@@ -244,405 +214,233 @@ export const FileTree = React.memo(function FileTreeComponent({
...
-                payload: { uuid: pipelineUuid, path: cleanFilePath(newPath) },
-              });
-            }
+      const isRename =
+        moves.length === 1 &&
+        deriveParentPath(moves[0][0]) === deriveParentPath(moves[0][1]);
 
-            return change;
-          })
+      if (overwrites.length) {
```
[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx:337]: I'll see what I can do here. I think I need to come up with some pattern here, the different dialogs interact with each other in interesting ways in general. For instance, if you're breaking some files _and_ you're overwriting others, both dialogs should be shown, and you should be able to back out of either one of them.

[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/pipeline-view/file-manager/FileTree.tsx]: OK, figured out a better pattern for this based on promises. 👍 

Have a look at [8041bb0](https://github.com/orchest/orchest/pull/1115/commits/8041bb024cc2861e3d314b4efca67ad8f4fcf278)

**Comment by mausworks** [comment]
https://github.com/orchest/orchest/pull/1115#issuecomment-1185548910

I just removed this dialog for now. Editors generally behave this way.

**Review by iannbing — APPROVED** [review]

**Pull request merged**